Refactor trace segment handling to remove `TraceSegmentId::index()` and introduce `TraceShape<T>`

surajk-m opened the pull request:
Closes #442 

This PR addresses the issue raised in #438 by removing the `TraceSegmentId::index()` method and introducing a type safe approach to handling trace segments across the codebase. The goal is to eliminate the reliance on numeric indices for referencing trace segments (main, aux, and quotient) and unify the representation of trace data using the proposed TraceShape<T> and FullTraceShape<T> structs, as suggested by @adr1anh and @bobbinth.

@adr1anh @bobbinth , I would appreciate your feedback on whether the `TraceShape<T>` and `FullTraceShape<T>` implementations fully cover all use cases in the AIR and ACE modules.

surajk-m committed:
feat: add `TraceShape<T>`,`FullTraceShape<T>` and remove `TraceSegmentId::index()`

surajk-m committed:
refactor: constraints to use `TraceShape<T>` for boundary and integrity

surajk-m committed:
refactor: `MirToAir` pass to use `TraceShape<T>` for trace columns

surajk-m committed:
modify ACE codegen to access trace segments

adr1anh commented on air/src/ir/mod.rs:
```diff
@@ -22,6 +22,94 @@ pub use self::{
+/// A fixed two segment trace shape containing values for the main and aux segments.
+#[derive(Clone, Debug, PartialEq, Eq, Default)]
+pub struct TraceShape<T> {
+    main: T,
+    aux: T,
```
These could be `pub`, avoiding the need for the `get` and `get_mut` methods. The `Index` and `IndexMut` implementation would be used for indexing with a `TraceSegmentID`.

adr1anh commented on air/src/passes/translate_from_mir.rs:
```diff
@@ -634,12 +642,11 @@ impl AirBuilder<'_> {
+            boundary.span(),
+            trace_access.column,
+            boundary.kind,
+        ) {
```
Adding the following trait implementations could simplify this call. Up to you if you think it is worth it.

```
impl<T> core::ops::Index<usize> for TraceShape<T> {
    type Output = &T;

    fn index(&self, index: usize) -> &Self::Output {
        match index {
            0 => &self.main,
            1 => &self.aux,
            _ => panic!("invalid segment index"),
    }
```

surajk-m replied:
The proposed `Index<usize>` trait implementation for `TraceShape<T>` looks promising and could indeed simplify indexing calls. I'll integrate it with a matching `IndexMut<usize>` for mutable access.

adr1anh reviewed (commented):
Thanks for tackling this, it looks really good. 

I think we could go a bit further here with the following suggestions. I might be missing some details that make these points redundant, so up to you whether they are worth implementing. 
- Adding `impl Index{Mut}<usize> for {Full}TraceShape` would simplify some calls that still rely on indices. Panicking when encountering an out-of-range index would be okay as that matches indexing behavior of other types.
- Replace `trace_segments: [[InputRegion; 3]; 2]` in the ACE Layout with `[FullTraceShape; 2]`
- We can remove `get` and `get_mut` as these are handled by the `Index` trait.
- Make the fields  of `TraceShape` public.
- I think keeping `TraceSegmentId::index` could still be useful, though it could be replaced by `From<TraceSegmentId> for usize` since the conversion never fails.

bobbinth commented:
@surajk-m - are you still planning to finish this PR?

surajk-m commented:
Hey @bobbinth @adr1anh,

Thanks for the feedback! Sorry for the delayed reply. Yes, I’m still planning to finish this up soon. I’ve been tied up with other tasks, but I’m aiming to push the changes by tomorrow.

impl `Index{Mut}<usize> for {Full}TraceShape` will clean up the index based calls nicely. I’ll add impl `Index<usize>` and `IndexMut<usize>` for both `TraceShape` and `FullTraceShape`, with panics on out of range indices to match standard behavior. The example impl you provided for Index is spot on, I’ll use something similar and extend it for mutability.

I’ll keep index() for now since it’s explicit and useful in some contexts, but I’ll also add the From `impl` as a convenience (it’s infallible, so no issues there).

Regarding `trace_segments: [[InputRegion; 3]; 2]` in ACE Layout, switching to `[FullTraceShape; 2]` makes total sense for consistency and better type safety. I’ll make that change.

surajk-m committed:
feat: implement `Index/From` traits for `TraceShape` and simplify struct access

surajk-m committed:
chore: changelog for issue #442

adr1anh commented:
This looks great! One final comment: how easy is it to use this in the `AceInputs`?

bobbinth commented:
@surajk-m - seems like the build is failing here? would you be able to fix this?

surajk-m committed:
refactor(ace): switch to `FullTraceShape` and update usages

- Change `AceVars.segments` to `[FullTraceShape<Vec<QuadFelt>>; 2]`
- Add `TraceShape::iter_mut()` and use array `into_iter()`
- Build `FullTraceShape` rows in `AirInputs::into_ace_vars`
- Simplify `to_memory_vec` loop by indexing via shapes directly
- Update `ACE` tests to construct `FullTraceShape` rows

surajk-m committed:
Update `TraceShape::iter_mut()` to return `impl Iterator<Item = (TraceSegmentId, &mut T)>`

surajk-m commented:
Hey @bitwalker @bobbinth, thanks for flagging the build issue,

> Super glad you like the changes! On the `AceInputs` no worries at all. Since we kept the `AirInputs` interface as is (still exposing `main`, `aux`, and `quotient` separately), all the updates to `AceVars.segments` and the indexing logic are nicely tucked away. We don’t need to tweak anything in existing `AceInputs` code. 

> The switch to `[FullTraceShape<Vec<QuadFelt>>; 2]` makes segment handling smoother, and `into_ace_vars` now uses `FullTraceShape::new(main_*, aux_*, quotient_*)` for more consistent row building. 

> The memory write loop’s also cleaner with the new `Index/IndexMut<usize>` setup, using `layout.trace_segments[row][i]` and 
`self.segments[row][i]`. It all plays nicely with `AceInputs` out of the box.

surajk-m committed:
rustfmt `CI` checks

Leo-Besancon commented on air/src/ir/mod.rs:
```diff
@@ -24,6 +24,134 @@ pub use self::{
+    }
+}
+
+impl<T> core::ops::Index<usize> for TraceShape<T> {
```
It seems that these two implementations are never used in the codebase (it's only used for `FullTraceShape` I believe)

```rust
impl<T> core::ops::Index<usize> for TraceShape<T>
impl<T> core::ops::IndexMut<usize> for TraceShape<T>
```

I think we could remove them as we have `Index<TraceSegmentId>` and `IndexMut<TraceSegmentId>`, but we could also keep them to stay consistent with `FullTraceShape` (that does rely on them).

What do you think @adr1anh?

adr1anh replied:
I'd favor keeping them for consistency :)

Leo-Besancon commented on air/src/passes/translate_from_mir.rs:
```diff
@@ -65,19 +71,21 @@ impl Pass for MirToAir<'_> {
                 bus_bindings_map.insert(binding.name.unwrap(), binding.offset);
             }
-            trace_columns.insert(aux_trace_segment.id, aux_trace_segment);
+            trace_columns_aux = aux_trace_segment;
```
If there are buses, you currently create two TraceSegment for the aux trace (one empty to initialize it, that is replaced with one with the bus bindings).

I think it would be cleaner to only allocate it once:
```rust
let trace_columns_aux = if buses.empty() {
  // return empty aux TraceSegment
} else {
  // return TraceSegment with bus_raw_bindings  
}
```

surajk-m replied:
Got it, I’ve updated it so the aux trace is allocated only once.

Leo-Besancon commented on codegen/ace/src/tests/quotient.rs:
```diff
@@ -66,7 +66,10 @@ pub fn eval_quotient(air: &Air, ace_vars: &AceVars, log_trace_len: u32) -> QuadF
+                        TraceSegmentId::Main => 0,
+                        TraceSegmentId::Aux => 1,
+                    };
                     ace_vars.segments[access.row_offset][segment_index][access.column]
```
You can use directly 
```rust
ace_vars.segments[access.row_offset][access.segment][access.column]
```

Leo-Besancon reviewed (commented):
Good job, it's a lot cleaner with this PR!

I've left a few (non-blocking on my side) comments, but otherwise it looks good to me.

adr1anh commented:
I think it would still be good to use `FullTraceShape` in [`AirInputs`](https://github.com/surajk-m/air-script/blob/57e00abab91be07b047e34f5faad6d5d243236a4/codegen/ace/src/inputs.rs#L24) to replace `main, aux, quotient`.

surajk-m committed:
use `FullTraceShape` in `AirInputs` and make aux trace allocation single pass

surajk-m commented:
Hey @adr1anh, `AirInputs` now uses `FullTraceShape`. Thanks for the feedback above @adr1anh @Leo-Besancon @bobbinth, we can merge the PR.

Leo-Besancon reviewed (approved):
LGTM!

bobbinth commented:
@surajk-m - could you please merge the latest `next` into this branch? The test should pass after this.

bobbinth reviewed (approved):
Looks good! Thank you!
